### PR TITLE
`Lectures`: Mark units as completed when downloading a merged pdf

### LINKED ISSUE #7865: `Lectures`: Downloading complete pdf does not complete lecture units
### Describe the bug

When downloading the complete PDF of a lecture, the lecture units are not marked as complete.
This is a different behavior compared to downloading a single lecture unit file.

### To Reproduce

Requirements:
- 1 Student
- Course with Lecture and multiple Lecture Units containing PDFs.

1. Go to Lecture
2. Click on 'Download complete PDF'
None of the lecture units are marked as complete.

### Expected behavior

The lecture units associated with the downloaded PDFs should be marked as complete.

### Screenshots

_No response_

### Which version of Artemis are you seeing the problem on?

6.7.4

### What browsers are you seeing the problem on?

Chrome

### Additional context

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by Strohgelaender**
#### General
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).

### Motivation and Context
Fixes #7865 

### Description
We now mark all units, that are included into the merged pdf, as completed.

### Steps for Testing
Prerequisites:
- 2 Student
- 1 Lecture with pdf lecture units

1. Open the lecture page
2. Click on merge pdf
3. Check that all lecture units are marked as completed.
4. Check that the student can manually mark the unit as un-completed.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

<details>

## Summary by CodeRabbit

- **New Features**
	- Introduced a method to track completion status for multiple lecture units, enhancing the learning progress tracking for users.
- **Refactor**
	- Improved code formatting and structure in repository and service layers for better readability.
	- Enhanced file handling logic in the `mergePdfFiles` method for improved performance and reliability.

</details>

### CODE REVIEW

**Review by MaximilianAnzinger — APPROVED** [review]
Thx for this quick bugfix and the code improvements 👍

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]
Please adjust the client tests accordingly.

**Review by MaximilianJG — APPROVED** [review]
Tested on TS4. Works as expected: When I "download complete PDF" all lecture pdfs get marked as completed. 
Screenshot 2024-01-09 at 20 33 37

**Review by MaximilianAnzinger — DISMISSED** [review]
Thank you for fixing the tests.

Failing E2E tests are unrelated 👍

**Review by egekurt123 — APPROVED** [review]
Manually tested on ts5 with student and admin roles, works as expected. 

However, after downloading merged pdf, you need to refresh the page to see that units are marked as complete. While downloading the units one by one, this is not the case

**Review by farisd16 — DISMISSED** [review]
Tested on ts5, I also had to refresh the page in order to see that the units are marked.

**Review by xHadie — APPROVED** [review]
_Manually tested on TS2_LEGACY_
Works as expected, ignoring the oddity that @egekurt123 mentioned.
I expected to see the boxes checked without a refresh.

**Review by Gusti2010 — DISMISSED** [review]
Tested manually on ts11_Legacy together with @konrad2002. Works as expected. However, you have to refresh the page after downloading the complete pdf
image

**Review by konrad2002 — APPROVED** [review]
manually tested on legacy TS11 with @Gusti2010 , after downloading the complete pdf and refreshing the page/reopen the unit, all files are marked as completed

**Review by coolchock — DISMISSED** [review]
tested on ts5, works as expected. The page doesn't have to be reloaded anymore. Code lgtm.

**Review by MaximilianAnzinger — DISMISSED** [review]

**Review by egekurt123 — APPROVED** [review]
Manually tested on ts5 again, now there is no need to reload the page to see boxes are marked as completed.

**Review by LenaKahle — APPROVED** [review]
Approve, manually tested on legacy ts11.

**Review by yassinsws — APPROVED** [review]
Code lgtm

**Review by MaximilianAnzinger — APPROVED** [review]

**Review by Predixx — APPROVED** [review]
Tested on ts2. LGTM

**Review by RY997 — DISMISSED** [review]
Code lgtm

**Review by georgyia — DISMISSED** [review]
tested manually on ts3, works as expected

**Review by MaximilianAnzinger — APPROVED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]
Failing GH tests are unrelated.

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java [hunk 1]
```diff
@@ -357,7 +360,7 @@ public ResponseEntity<byte[]> getLecturePdfAttachmentsMerged(@PathVariable Long
 
         List<String> attachmentLinks = lectureAttachments.stream()
                 .filter(unit -> authCheckService.isAllowedToSeeLectureUnit(unit, user) && "pdf".equals(StringUtils.substringAfterLast(unit.getAttachment().getLink(), ".")))
-                .map(unit -> FilePathService.getAttachmentUnitFilePath()
+                .peek(unit -> lectureUnitService.setLectureUnitCompletion(unit, user, true)).map(unit -> FilePathService.getAttachmentUnitFilePath()
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:363]: This looks like we create on additional DB call per lecture unit. Would it be a good idea to set these completion states in a single DB call for all units at once?

**Review by bassner — DISMISSED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/LectureUnitService.java [hunk 2]
```diff
@@ -70,6 +65,49 @@ public void setLectureUnitCompletion(@NotNull LectureUnit lectureUnit, @NotNull
...
 
+    /**
+     * Set the completion status of all passed lecture units for the give user
+     * If the user completed the unit and completion status already exists, nothing happens
+     *
+     * @param lectureUnits List of all lecture units for which to set the completion flag
+     * @param user         The user that completed/uncompleted the lecture unit
+     * @param completed    True if the lecture unit was completed, false otherwise
+     */
+    public void completeAllLectureUnits(List<? extends LectureUnit> lectureUnits, @NotNull User user, boolean completed) {
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/LectureUnitService.java:76]: Right now the method name suggest that you set complete = true for all units
```suggestion
    public void setCompletedForAllLectureUnits(List<? extends LectureUnit> lectureUnits, @NotNull User user, boolean completed) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java [hunk 3]
```diff
@@ -353,14 +356,14 @@ public ResponseEntity<byte[]> getLecturePdfAttachmentsMerged(@PathVariable Long
...
-        List<String> attachmentLinks = lectureAttachments.stream()
+        List<AttachmentUnit> lectureAttachments = attachmentUnitRepository.findAllByLectureIdAndAttachmentTypeElseThrow(lectureId, AttachmentType.FILE).stream()
                 .filter(unit -> authCheckService.isAllowedToSeeLectureUnit(unit, user) && "pdf".equals(StringUtils.substringAfterLast(unit.getAttachment().getLink(), ".")))
-                .map(unit -> FilePathService.getAttachmentUnitFilePath()
-                        .resolve(Path.of(String.valueOf(unit.getId()), StringUtils.substringAfterLast(unit.getAttachment().getLink(), "/"))).toString())
                 .toList();
 
+        lectureUnitService.completeAllLectureUnits(lectureAttachments, user, true);
+        List<String> attachmentLinks = lectureAttachments.stream().map(unit -> FilePathService.getAttachmentUnitFilePath()
+                .resolve(Path.of(String.valueOf(unit.getId()), StringUtils.substringAfterLast(unit.getAttachment().getLink(), "/"))).toString()).toList();
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:365]: Not in scope but maybe you can fix this anyway here:
We should fetch the actual links and not reconstruct it. This is prone to errors if the path generation changes.

**Review by julian-christl — APPROVED** [review]
Thanks for implementing my suggestions.

Code looks good to me.

**Review by georgyia — DISMISSED** [review]
manually tested on ts2, works as expected

**Review by tobias-lippert — APPROVED** [review]
Code LGTM

**Review by MaximilianJG — APPROVED** [review]

**Review by AntonGeTUM — APPROVED** [review]
Manually tested on ts1, works as described

**Review by MaximilianAnzinger — APPROVED** [review]

**Review by bassner — APPROVED** [review]
LGTM

**Review by MaximilianAnzinger — APPROVED** [review]

**Pull request merged**